daemon/daemon_main.go

```diff
@@ -1118,6 +1107,42 @@ func runDaemon() {
 	d.initK8sSubsystem()
 	bootstrapStats.k8sInit.End(true)
 
+	// If K8s is enabled we can do the service translation automagically by
+	// looking at services from k8s and retrieve the service IP from that.
+	// This makes cilium to not depend on kube dns to interact with etcd
+	var goopts *kvstore.ExtraOptions
+	if k8s.IsEnabled() {
+		goopts = &kvstore.ExtraOptions{}
+		log := log.WithField(logfields.LogSubsys, "etcd")
+		goopts.DialOption = []grpc.DialOption{
+			grpc.WithDialer(func(s string, duration time.Duration) (conn net.Conn, e error) {
```

Comment:
Please add a comment that describes how this works. In particular:
 - How does retry work?
 - What happens when the service is not populated yet?
 - What is the retry interval?
 - What happens when the backend IP disappears?